src/backend/optimizer/path/allpaths.c

```diff
@@ -2033,6 +2033,15 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 	 */
 	if (!root->config->gp_cte_sharing || cte->cterefcount == 1)
 	{
+	    /*
+	     * If plan sharing is disabled, we avoid performing DML inside CTE
+	     * for each reference. It'll cause duplicated DML operations or
+	     * mutation errors during cdbparallelize().
+	     */
+	    if (cte->cterefcount > 1 &&
```

Comment:
1) Currently, this block use spaces for indentation.
2) Why do we place this block inside `if (!root->config->gp_cte_sharing || cte->cterefcount == 1)` condition?

Reply:
2. It is placed here in order to separate 2 cases: when **gp_cte_sharing is on** and when it is **off**. At the first case, the plan is created, and, the error occurs only at execution step. 
```
create table with_dml (i int, j int) distributed by (i);
explain (costs off) with cte as (
    insert into with_dml select i, i * 100 from generate_series(1,5) i
    returning i
) select count(*) from cte where i < (select avg(i) from cte);
ERROR:  Material operator does not support projection (nodeMaterial.c:358)
```
The plan tree is built and has approximately the following structure. (Explain fails after planning, and plan is invalid, but not because of inlining. Therefore, we don't put the error for this case).
```
                                QUERY PLAN                                   
--------------------------------------------------------------------------------
 Aggregate
   InitPlan 1 (returns $1)  (slice5)
     ->  Aggregate
           ->  Gather Motion 3:1  (slice4; segments: 3)
                 ->  Aggregate
                       ->  Insert on with_dml with_dml_1
                             ->  Redistribute Motion 1:3  (slice3; segments: 1)
                                   Hash Key: i_1.i
                                   ->  Function Scan on generate_series i_1
      ->  Gather Motion 3:1  
         ->  Aggregate
               ->  Subquery Scan on cte
                     Filter: ((cte.i)::numeric < $1)
                     -> Shared Scan
                          -> Materialize
                                 ->  Insert on with_dml
                                       ->  Redistribute Motion 1:3  (slice1; segments: 1)
                                               Hash Key: i.i
                                               ->  Function Scan on generate_series i
```
And for this case we let planner work further, because the cte sharing has its own logic.

Reply:
1. Replace spaces with tabs?

Reply:
> 2\. It is placed here in order to separate 2 cases: when **gp_cte_sharing is on** and when it is **off**. At the first case, the plan is created, and, the error occurs only at execution step.

Still can't understand how it helps us. Our query now gives the same error regardless of `gp_cte_sharing` GUC is on or not. In our case the value of `root->config->gp_cte_sharing` is always false because of `subselect.c:652`

Do we have some cases which throw an error for nothing if new check placed like in older version (not under if condition)?

Reply:
> 1. Replace spaces with tabs?

You need to run `pgindent` and format this block according to its output.

Reply:
Actually the `root->config->gp_cte_sharing` is set to `true` for our case, because cte is planned twice (what is wrong imho): First time is when we call preprocess_qual_conditions [from](https://github.com/arenadata/gpdb/blob/2de3b73a03045062849da9b99612bcd022334a7b/src/backend/optimizer/plan/planner.c#L731) `subquery_planner`. Qual for our case is `i < SubLink`, where the SubLink is for (select avg(i) from cte). Further, in `make_subplan` the `gp_cte_sharing` is set to false, and then cte is [planned](https://github.com/arenadata/gpdb/blob/2de3b73a03045062849da9b99612bcd022334a7b/src/backend/optimizer/plan/subselect.c#L692) as `initPlan`. Second time is during planning main SELECT and processing the `joinlist`. And for this time the `root->config->gp_cte_sharing` is set to `true`, and we get in main plan the part with shared scan.

So, yep, for queries like that there is no difference for the place where to put the if-clause.
However for the queries like that
```
explain (costs off) with cte as (
    insert into with_dml select i, i * 100 from generate_series(1,5) i
    returning i
) select 1 from cte join (select * from cte) as tmp using (i);
```
The cte is also planned twice, but  the `root->config->gp_cte_sharing` is set to `true` in both cases. Therefore, I suppose, the concept of plan sharing should be considered separately from our case and It would be better to leave it as it's now.

Reply:
Have run the `pgindent`

Reply:
> The cte is also planned twice, but the `root->config->gp_cte_sharing` is set to `true` in both cases. Therefore, I suppose, the concept of plan sharing should be considered separately from our case and It would be better to leave everything as it's now.

Do we have a separate ticket to fix gp_cte_sharing behavior? I think we should somehow highlight this.

Reply:
No, we don' such ticket, except general 2483 about inlining. And the problem is mostly in that the planner processes the CTE twice regardless from gp_cte_sharing. It is general behaviour for all CTEs. And for this 3881 ticket I assumed that gp_cte_sharing is out of our consideration, because it potentially can fix the problem in terms of shared plans (however it will be too specific and the general issue with modifying cte will not be solved).

Reply:
All questions were answered. I'm going to approve this.